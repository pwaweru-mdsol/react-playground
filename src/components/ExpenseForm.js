import React, { Component } from "react";
import { SingleDatePicker } from "react-dates";
import "react-dates/initialize";
import moment from "moment";

const ERROR_MESSAGE = "Please enter a description or an amount";

export class ExpenseForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '' ,
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : '',
      calenderFocused: false,
      error: undefined
    }
  }

  onDescriptionChange = event => {
    const description = event.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = event => {
    const note = event.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = event => {
    const amount = event.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt: createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calenderFocused: focused }));
  };

  onSubmit = event => {
    event.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: ERROR_MESSAGE }));
    } else {
      this.setState(() => ({ error: undefined }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error ? <p>{ERROR_MESSAGE}!</p> : null}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="number"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />

          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calenderFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note for your exepnse (optional)"
            onChange={this.onNoteChange}
          />
          <button>Submit Expense</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
