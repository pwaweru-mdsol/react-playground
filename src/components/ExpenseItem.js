import React from "react";
import { Link } from "react-router-dom";
import moment from 'moment';
import numeral from 'numeral';

export const ExpenseItem = ({ uuid, description, amount, createdAt }) => (
  <div>
    <Link to={`/edit/${uuid}`}>
      <h3>{description}</h3>
    </Link>
    <p>{numeral(amount / 100).format('$0,0.00')}</p>
    <p>{moment(createdAt).format('MMMM Do, YYYY')}</p>
  </div>
);

export default ExpenseItem;
