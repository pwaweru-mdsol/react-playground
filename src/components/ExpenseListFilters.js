import React, { Component } from "react";
import { connect } from "react-redux";
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from "../actions/filters";

import uuid from "uuid/v4";

import { DateRangePicker } from "react-dates";

export class ExpenseListFilters extends Component {
  state = {
    calenderFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = calenderFocused => {
    this.setState(() => ({ calenderFocused }));
  };

  onTextChange = event => {
    this.props.setTextFilter(event.target.value);
  }

  onSortChange = event => {
    if(event.target.value == 'date') {
      this.props.sortByDate()
    } else if(event.target.value == 'amount') {
      this.props.sortByAmount()
    }
  }

  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />

        <select onChange={this.onSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>

        <DateRangePicker
          startDateId={uuid()}
          endDateId={uuid()}
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calenderFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          showClearDates={true}
          isOutsideRange={() => false}
        />
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    filters: state.filters
  };
};

const mapDispatchToPropes = dispatch => ({
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: setEndDate => dispatch(setEndDate(setEndDate)),
});

export default connect(mapStateToProps, mapDispatchToPropes)(ExpenseListFilters);
