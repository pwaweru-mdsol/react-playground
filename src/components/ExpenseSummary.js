import React from "react";
import { connect } from "react-redux";

export const ExpenseSummary = () => (
  <div>
    <p>Expense Summary</p>
  </div>
);

export default connect()(ExpenseSummary);