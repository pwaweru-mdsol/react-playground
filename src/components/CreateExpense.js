import React from "react";

import ExpenseForm from "./ExpenseForm";

const CreateExpense = () => (
  <div>
    Create Expense!
    <ExpenseForm />
  </div>
);

export default CreateExpense;
