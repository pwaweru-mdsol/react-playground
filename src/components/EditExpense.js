import React from "react";
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense } from "../actions/expenses";

const EditExpense = props => {
  return (
  	<div>
  		<ExpenseForm 
	  		expense={props.expense}
	  		onSubmit={(expense) => {
	  			props.dispatch(editExpense(props.expense.uuid, expense));
	  			props.history.push('/');
  			}}
  		/>
  	</div>);
};

const mapStateToProps = (state, props) => {
	return {
		expense: state.expenses.find(expense => expense.uuid === props.match.params.uuid)
	}
}

export default connect(mapStateToProps)(EditExpense);
