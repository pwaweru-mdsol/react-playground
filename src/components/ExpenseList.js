import React from 'react';
import { connect } from 'react-redux';

import ExpenseItem from './ExpenseItem';

const ExpenseList = (props) => (
	<div>
		<h1>Expense List</h1>
		{
			props.expenses.map(expense => (
				<ExpenseItem description={expense.description} amount={expense.amount} createdAt={expense.createdAt}/>
			))
		}
	</div>
);

const mapStateToProps = (state => {
	return { expenses: state.expenses, filters: state.filters }
});

export default connect(mapStateToProps)(ExpenseList);		