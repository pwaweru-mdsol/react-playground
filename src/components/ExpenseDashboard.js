import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpenseSummary from './ExpenseSummary';

const ExpenseDashboard = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseSummary />
    <ExpenseList />
  </div>
);

export default ExpenseDashboard;
