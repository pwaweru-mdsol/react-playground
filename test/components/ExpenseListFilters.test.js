import React from 'react'
import { shallow } from 'enzyme';
import { ExpenseListFilters } from '../../src/components/ExpenseListFilters';
import { defaultFilters, filters } from '../fixtures/filters';
import moment from 'moment'

let setTextFilterSpy, sortByDateSpy, sortByAmountSpy, setStartDateSpy, setEndDateSpy, wrapper;

describe('ExpenseListFilters', () => {
  beforeEach(() => {
      setTextFilterSpy = jest.fn();
      sortByDateSpy = jest.fn();
      sortByAmountSpy = jest.fn();
      setStartDateSpy = jest.fn();
      setEndDateSpy = jest.fn();
      wrapper = shallow(<ExpenseListFilters 
          filters={defaultFilters}
          setTextFilter={setTextFilterSpy}
          sortByDate={sortByDateSpy}
          sortByAmount={sortByAmountSpy}
          setStartDate={setStartDateSpy}
          setEndDate={setEndDateSpy}
      /> )
  })

  test('should render ExpenseListFilters', () => {
      expect(wrapper).toMatchSnapshot();
  });

  test('should render ExpenseListFilters with updated filters', () => {
      wrapper.setProps({
          filters: filters
      })
      expect(wrapper).toMatchSnapshot();
  });

  test('should handle onTextChange', () => {
      const value = 'hello'
      wrapper.find('input').simulate('change', {
          target: { value }
      });
      expect(setTextFilterSpy).toHaveBeenCalledWith(value);
  });

  test('should handle sortByDate', () => {
      const value = 'date'
      wrapper.find('select').simulate('change', {
          target: { value }
      });
      expect(sortByDateSpy).toHaveBeenCalled();
  });

  test('should handle sortByAmount', () => {
      const value = 'amount'
      wrapper.find('select').simulate('change', {
          target: { value }
      })
      expect(sortByAmountSpy).toHaveBeenCalled();
  });

  test('should handle onDatesChange', () => {
      const startDate = moment(0).add(5, 'years')
      const endDate = moment(0).add(10, 'years');
      wrapper.find('withStyles(DateRangePicker)').prop('onDatesChange')({ startDate, endDate });
      expect(setStartDateSpy).toHaveBeenLastCalledWith(startDate);
      expect(setEndDateSpy).toHaveBeenLastCalledWith(endDate);
  });

  test('should handle onFocusChange startDate', () => {
      const calenderFocused = 'startDate'
      wrapper.find('withStyles(DateRangePicker)').prop('onFocusChange')(calenderFocused);
      expect(wrapper.state('calenderFocused')).toBe(calenderFocused)
  });
});