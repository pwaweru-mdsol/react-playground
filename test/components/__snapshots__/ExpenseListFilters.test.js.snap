// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseListFilters should render ExpenseListFilters 1`] = `
<div>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
  <select
    onChange={[Function]}
  >
    <option
      value="date"
    >
      Date
    </option>
    <option
      value="amount"
    >
      Amount
    </option>
  </select>
  <withStyles(DateRangePicker)
    anchorDirection="left"
    appendToBody={false}
    block={false}
    calendarInfoPosition="bottom"
    customArrowIcon={null}
    customCloseIcon={null}
    customInputIcon={null}
    daySize={39}
    disableScroll={false}
    disabled={false}
    displayFormat={[Function]}
    enableOutsideDays={false}
    endDate={null}
    endDateId="2c65e4c2-d555-4085-9650-8dc33148f0d4"
    endDatePlaceholderText="End Date"
    firstDayOfWeek={null}
    focusedInput={null}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    initialVisibleMonth={null}
    inputIconPosition="before"
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    minimumNights={1}
    monthFormat="MMMM YYYY"
    navNext={null}
    navPrev={null}
    noBorder={false}
    numberOfMonths={1}
    onClose={[Function]}
    onDatesChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    openDirection="down"
    orientation="horizontal"
    phrases={
      Object {
        "calendarLabel": "Calendar",
        "chooseAvailableEndDate": [Function],
        "chooseAvailableStartDate": [Function],
        "clearDates": "Clear Dates",
        "closeDatePicker": "Close",
        "dateIsSelected": [Function],
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "focusStartDate": "Interact with the calendar and add the check-in date for your trip.",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month.",
        "jumpToPrevMonth": "Move backward to switch to the previous month.",
        "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
        "moveFocusByOneMonth": "Switch months.",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
        "openThisPanel": "Open this panel.",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field.",
        "selectFocusedDate": "Select the date in focus.",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonth={null}
    reopenPickerOnClearDates={false}
    required={false}
    screenReaderInputMessage=""
    showClearDates={true}
    showDefaultInputIcon={false}
    small={false}
    startDate={null}
    startDateId="3f767b92-ac8d-4ef3-b1dd-aaa08f0979d0"
    startDatePlaceholderText="Start Date"
    verticalHeight={null}
    verticalSpacing={22}
    weekDayFormat="dd"
    withFullScreenPortal={false}
    withPortal={false}
  />
</div>
`;

exports[`ExpenseListFilters should render ExpenseListFilters with updated filters 1`] = `
<div>
  <input
    onChange={[Function]}
    type="text"
    value="Hello"
  />
  <select
    onChange={[Function]}
  >
    <option
      value="date"
    >
      Date
    </option>
    <option
      value="amount"
    >
      Amount
    </option>
  </select>
  <withStyles(DateRangePicker)
    anchorDirection="left"
    appendToBody={false}
    block={false}
    calendarInfoPosition="bottom"
    customArrowIcon={null}
    customCloseIcon={null}
    customInputIcon={null}
    daySize={39}
    disableScroll={false}
    disabled={false}
    displayFormat={[Function]}
    enableOutsideDays={false}
    endDate={"1970-01-04T00:00:00.000Z"}
    endDateId="f9d6afd3-a56c-4738-a0ac-44f39bc8de41"
    endDatePlaceholderText="End Date"
    firstDayOfWeek={null}
    focusedInput={null}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    initialVisibleMonth={null}
    inputIconPosition="before"
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    minimumNights={1}
    monthFormat="MMMM YYYY"
    navNext={null}
    navPrev={null}
    noBorder={false}
    numberOfMonths={1}
    onClose={[Function]}
    onDatesChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    openDirection="down"
    orientation="horizontal"
    phrases={
      Object {
        "calendarLabel": "Calendar",
        "chooseAvailableEndDate": [Function],
        "chooseAvailableStartDate": [Function],
        "clearDates": "Clear Dates",
        "closeDatePicker": "Close",
        "dateIsSelected": [Function],
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "focusStartDate": "Interact with the calendar and add the check-in date for your trip.",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month.",
        "jumpToPrevMonth": "Move backward to switch to the previous month.",
        "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
        "moveFocusByOneMonth": "Switch months.",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
        "openThisPanel": "Open this panel.",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field.",
        "selectFocusedDate": "Select the date in focus.",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonth={null}
    reopenPickerOnClearDates={false}
    required={false}
    screenReaderInputMessage=""
    showClearDates={true}
    showDefaultInputIcon={false}
    small={false}
    startDate={"1970-01-01T00:00:00.000Z"}
    startDateId="279bf3a5-5779-4998-81a5-a89d08280b36"
    startDatePlaceholderText="Start Date"
    verticalHeight={null}
    verticalSpacing={22}
    weekDayFormat="dd"
    withFullScreenPortal={false}
    withPortal={false}
  />
</div>
`;
